test/server/server_test.go

```diff
@@ -0,0 +1,331 @@
+//go:build integration
+// +build integration
+
+/*
+Copyright 2021 Andy Dalton
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package server_test
```

Comment:
Could we add a test for correct authorization? Ie, user trying to interact with someone else's job. I'd toss in an admin test as well.

Reply:
This is an excellent example of why unit tests that I've included (along with the associated mocks) are valuable.  This is a check for a specific, well-defined behavior implemented by the `jobmanagerServer` component that we can test in isolation.  The only thing running it through the full gRPC stack does beyond what the unit tests (and other existing integration tests) do is verify that the test certificates really do have different CommonNames (and that the admin one actually does have a CommonName of "administrator")

I've added both the unit test and the integration test.  Note that the integration tests are significantly slower.

```none
// === RUN   Test_jobmanagerServer_Multitenant
// --- PASS: Test_jobmanagerServer_Multitenant (0.00s)

vs 

// === RUN   Test_clientServer_Multitenant
// --- PASS: Test_clientServer_Multitenant (1.00s)
```

and

```none
// === RUN   Test_jobmanagerServer_AdministratorCanSeeAllJobs
// --- PASS: Test_jobmanagerServer_AdministratorCanSeeAllJobs (0.00s)
//
// vs
//
// === RUN   Test_clientServer_AdministratorCanSeeAllJobs
// --- PASS: Test_clientServer_AdministratorCanSeeAllJobs (0.84s)
```

One-second tests have the potential to build up very quickly over time.

I currently have 64 unit test across the entire project, and it takes less time to run all of them than it does to run one of these tests.

```none
DONE 64 tests in 0.597s
```

I'm not arguing that we shouldn't include integration tests, but I'd be very selective about the tests that I'd include and the cadence at which I run them.

Reply:
Ack.